R/teal_lockfile.R

```diff
@@ -23,92 +44,154 @@
+        {
+          options(opts)
+          do.call(Sys.setenv, sysenv)
+          .libPaths(libpaths)
+          setwd(wd)
+
+          run(lockfile_path = lockfile_path)
+        },
+        run = renv_snapshot,
+        lockfile_path = lockfile_path,
+        opts = options(),
+        libpaths = .libPaths(),
+        sysenv = as.list(Sys.getenv()),
+        wd = getwd()
       )
-    )
-    if (any(grepl("Lockfile written", renv_logs))) {
-      logger::log_trace("lockfile created successfully.")
-    } else {
-      logger::log_trace("lockfile created with issues.")
     }
+  )
 
-    lockfile_path
+  suppressWarnings({ # 'package:stats' may not be available when loading
+    process$invoke()
   })
-  if (close) {
-    # If the initial backend was only sequential, bring it back.
-    promises::then(promise, onFulfilled = function() {
-      future::plan(future::sequential)
-    })
-  }
-  promise
+
+  logger::log_trace("Lockfile creation started based on { getwd() }.")
+
+  process
 }
 
+#' @rdname teal_lockfile
+#' @keywords internal
+renv_snapshot <- function(lockfile_path) {
+  checkmate::assert_string(lockfile_path)
+
+  out <- utils::capture.output(
+    res <- renv::snapshot(
+      lockfile = lockfile_path,
+      prompt = FALSE,
+      force = TRUE
+      # type is taken from renv::settings$snapshot.type()
+    )
+  )
+
+  list(
+    out = out,
+    res = res,
+    path = lockfile_path
+  )
+}
+
+#' @rdname teal_lockfile
+#' @keywords internal
+teal_lockfile_process_tracker <- function(process) {
+  checkmate::assert_class(process, "ExtendedTask")
+
+  observeEvent(process$status(), {
+    if (process$status() == "initial" || process$status() == "running") {
+      shinyjs::html("teal-lockFileStatus", "Creating lockfile...")
+    } else if (process$status() == "success") {
+      result <- process$result()
+      if (any(grepl("Lockfile written to", result$out))) {
+        logger::log_trace("Lockfile {result$path} containing { length(result$res$Packages) } packages created.")
+        if (any(grepl("WARNING:", result$out)) || any(grepl("ERROR:", result$out))) {
+          logger::log_warn("Lockfile created with warning(s) or error(s):")
+          for (i in result$out) {
+            logger::log_warn(i)
+          }
+        }
+
+        shinyjs::html("teal-lockFileStatus", "Application lockfile ready.")
+        shinyjs::hide("teal-lockFileStatus", anim = TRUE)
+
+        shinyjs::enable("teal-lockFileLink")
+      } else {
+        warning("Lockfile creation failed.")
+        shinyjs::html("teal-lockFileStatus", "Lockfile creation failed.")
+        shinyjs::disable("teal-lockFileLink")
+      }
+    } else if (process$status() == "error") {
+      warning("Lockfile creation failed.")
+      shinyjs::html("teal-lockFileStatus", "Lockfile creation failed.")
+      shinyjs::disable("teal-lockFileLink")
+    }
+  })
+}
+
+#' @rdname teal_lockfile
+#' @keywords internal
 teal_lockfile_downloadhandler <- function() {
```

Comment:
IMO separation of shiny namespaces is hard to maintain and since we do `disable(teal-lockFileLink)` and `shinyjs::html("teal-lockFileStatus)` we are risking to break this with any changes in the namespace in teal. We would need to be very careful with this in a future.

I think if we pass `lockfile_process` to the module we would be able to encapsulate everything in a single shiny namespace and we wouldn't have process split into multiple parts. 
```suggestion
ui_teal_lockfile <- function(id) {...}
srv_teal_lockfile <- function(id, process) {
  moduleServer(id, function(input, output, session) {
      if (!is.null(process)) {
        teal_lockfile_process_tracker(process)
      }
      
      ...
  })
}
```

The only problem is how to handover `lockfile_process` - either by argument `srv_teal_with_splash(..., process)` -> `srv_teal(..., process)` or by `session$userData$lockfile_process`. I prefer `userData` as `srv_teal` will be exported in the next release and we don't want to complicate public api.